lib/Dialect/SPIRV/Transforms/DecorateSPIRVCompositeTypeLayoutPass.cpp

```diff
@@ -0,0 +1,316 @@
+/// alignment of the array or matrix as defined above.
+/// 3. The ArrayStride, MatrixStride, and Offset decorations must be large
+/// enough to hold the size of the objects they affect (that is, specifying
+/// overlap is invalid).
+namespace {
+class VulkanLayoutUtils {
+public:
+  using Alignment = uint64_t;
+
+  /// Returns a new type with layout info. Assigns the type size in bytes to the
+  /// `size`. Assigns the type alignment in bytes to the `alignment`.
+  static Type decorateType(spirv::StructType structType,
+                           spirv::StructType::LayoutInfo &size,
+                           Alignment &alignment);
+  /// Checks whether a type is legal in terms of Vulkan layout info
+  /// decoration. A type is dynamically illegal if it's a composite type in the
+  /// StorageBuffer, PhysicalStorageBuffer, Uniform, and PushConstant Storage
+  /// Classes without layout informtation.
+  static bool isLegalType(Type type);
+
+private:
+  static Type decorateType(Type type, spirv::StructType::LayoutInfo &size,
+                           Alignment &alignment);
+  static Type decorateType(VectorType compositeType,
+                           spirv::StructType::LayoutInfo &size,
+                           Alignment &aligment);
+  static Type decorateType(spirv::ArrayType compositeType,
+                           spirv::StructType::LayoutInfo &size,
+                           Alignment &aligment);
+  /// Calculates the alignment for the given scalar type.
+  static Alignment getScalarTypeAlignment(Type scalarType);
+};
+
+Type VulkanLayoutUtils::decorateType(spirv::StructType structType,
+                                     spirv::StructType::LayoutInfo &size,
+                                     VulkanLayoutUtils::Alignment &alignment) {
+  if (structType.getNumElements() == 0) {
+    return structType;
+  }
+
+  llvm::SmallVector<Type, 4> memberTypes;
+  llvm::SmallVector<spirv::StructType::LayoutInfo, 4> layoutInfo;
+  llvm::SmallVector<spirv::StructType::MemberDecorationInfo, 4>
+      memberDecorations;
+
+  spirv::StructType::LayoutInfo structMemberOffset = 0;
+  VulkanLayoutUtils::Alignment maxMemberAlignment = 1;
+
+  for (uint32_t i = 0, e = structType.getNumElements(); i < e; ++i) {
+    spirv::StructType::LayoutInfo memberSize = 0;
+    VulkanLayoutUtils::Alignment memberAlignment = 1;
+
+    auto memberType = VulkanLayoutUtils::decorateType(
+        structType.getElementType(i), memberSize, memberAlignment);
+    structMemberOffset = llvm::alignTo(structMemberOffset, memberAlignment);
+    memberTypes.push_back(memberType);
+    layoutInfo.push_back(structMemberOffset);
+    // According to the Vulkan spec:
+    // "A structure has a base alignment equal to the largest base alignment of
+    // any of its members."
+    structMemberOffset += memberSize;
+    maxMemberAlignment = std::max(maxMemberAlignment, memberAlignment);
+  }
+
+  size = llvm::alignTo(structMemberOffset, maxMemberAlignment);
```

Comment:
Could you quote the spec here? "The Offset decoration of a member must not place it between the end of a structure or an array and the next multiple of the alignment of that structure or array."